### ISSUE TITLE
Short final window seems to hurt motion quality

### ISSUE THREAD

**Issue by scirbaludu** [Saganaki22/ComfyUI-Viggle-Animate-H3#8]
Hey, I tested the recent fix for the video ending issue and I think there is a small regression in the new chunk scheduling.

The good news is that the end of the video is no longer getting cut off.

But after the change that allows the last sampling window to be shorter, I noticed a pretty clear drop in motion quality right when the new part of the last chunk starts.

For example, with 192 frames at 24 fps, `chunk_frames = 124` and `overlap_frames = 22`, the current scheduler gives:

```
0-123
102-191
```

So the second window is only 90 frames.

In my test, around frame 124, which is about 5.17s, the source still has clear mouth movement, but the generated character basically stops moving its mouth there. The timing lines up almost exactly with the point where the second chunk starts generating new frames.

I tried keeping the new round-up behavior so the end of the source is still preserved, but making the last window full length again and shifting it backwards so it ends at the end of the video.

That gives:

```
0-123
68-191
```

With this, both windows stay at 124 frames and the motion issue disappeared for me. The end of the source video is still preserved too.

The change I tested in `plan_spans()` was basically this:

```python
O = _frames_to_latents(max(5, int(overlap_f)))
O = max(2, min(O, L - 5))
stride = L - O

starts = list(range(0, total_lat - L + 1, stride))

last = total_lat - L
if starts[-1] != last:
    starts.append(last)

return [
    (
        _frame_at_latent(s),
        _frame_at_latent(s + L) - 1,
        s,
        L,
    )
    for s in starts
]
```

And I kept this part from the new fix:

```python
total_f = _generation_frame_count(total_f)
```

So the tail is still rounded up instead of being truncated.

Looks like the short final window might be the part causing the quality regression, at least in this case.

Hope this is useful for the next round of code fixes

**Comment by scirbaludu** [comment]
update: I kept testing this and I think I found a much better way to handle the long-video continuation.

The larger overlap helped the first case, but on longer clips I still started getting motion drift around later chunks. I also tried independent chunks + stitching, but that caused visible jumps at the seams.

What ended up working much better was taking the same general idea as ComfyUI's `WanSCAILToVideo` `previous_frames` handling: use only a very small decoded frame anchor from the previous chunk instead of carrying a large raw latent overlap.

For Viggle/H3 I tested a fixed 5-frame anchor.

So instead of something like:

```text
chunk 1: 0-123
chunk 2: 68-191
chunk 3: 136-259
...
```

with a large latent carry, the normal step becomes 119 frames:

```text
chunk 1: 0-123
chunk 2: 119-242
...
```

The important part is that those 5 previous frames are not copied as raw diffusion latents.

After a chunk is generated, I decode the 5 frames that correspond to the start of the next window:

```python
next_a = windows[i + 1][0]
offset = next_a - a

decoded = vae.decode(out_v.to(dev))

anchor_frames = decoded[
    offset:offset + SCAIL_ANCHOR_FRAMES
]
```

Then for the next chunk those 5 pixel frames are encoded again with the H3 VAE:

```python
anchor_latent = vae.encode(anchor_frames)

carry = min(
    int(anchor_latent.shape[2]),
    latn,
)

v[:, :, :carry] = anchor_latent[:, :, :carry]
```

For H3, 5 frames map to 2 temporal latents:

```python
SCAIL_ANCHOR_FRAMES = 5

O = _frames_to_latents(SCAIL_ANCHOR_FRAMES)
stride = L - O
```

Only those anchor latents are pinned during sampling:

```python
if carry > 0:
    mask_v = torch.ones(
        [1, 1, latn, ch // 16, cw // 16],
        device=v.device,
    )

    mask_v[:, :, :carry] = 0.0
```

This made a pretty big difference in my tests. The motion stays aligned much better across later chunks and I no longer get the cumulative lip/motion desync I was seeing with the larger latent carry.

I also kept the full-size end-aligned final window from the previous fix:

```python
starts = list(
    range(0, total_lat - L + 1, stride)
)

last = total_lat - L

if starts[-1] != last:
    starts.append(last)
```

So the final sampling window is never shortened.

One detail here is that the final end-aligned window can have more than 5 frames of overlap.

For example, with a 322-frame source, H3 needs 328 frames internally because of the `17j+5` temporal grid:

```text
source: 322
generation extent: 328

chunk 1: 0-123
chunk 2: 119-242
chunk 3: 204-327
```

The last overlap there is larger because the final 124-frame window has been shifted back to end exactly at frame 327.

I still only use 5 frames as the actual continuation anchor. The rest of that overlap is just warm-up context and is not allowed to overwrite frames that were already accepted from the previous chunk:

```python
local_start = max(
    0,
    accepted_end - start,
)

global_start = start + local_start

copy_count = min(
    count - local_start,
    max(0, total_lat - global_start),
)

if copy_count > 0:
    master[
        :,
        :,
        global_start:global_start + copy_count,
    ] = video[
        :,
        :,
        local_start:local_start + copy_count,
    ]
```

The other small issue was output length.

I still think rounding UP before generation is the right thing to do:

```python
def _generation_frame_count(fc):
    fc = max(5, int(fc))
    return fc + (5 - fc) % 17
```

Otherwise real source frames can get dropped.

But after the final VAE decode, I now trim only the artificial grid padding back to the exact original source frame count:

```python
generated = int(images.shape[0])
source = int(chunks["source_frames"])

if generated < source:
    raise ValueError(
        "Final decode is shorter than the source"
    )

images = images[:source]
```

So for the example above:

```text
322 source frames
-> generate 328 frames on the valid H3 grid
-> final decode
-> trim 328 back to 322
```

No interpolation, duplication or resampling, just removing the 6 artificial tail frames.

With this version I'm finally getting frame-for-frame timing matching the driving video across the full clip, while also keeping the source ending and exact final frame count.

The main thing that seems to matter is:

```text
full 124-frame windows
+ 5 decoded/re-encoded previous frames
+ no large raw latent carry
+ full end-aligned final window
+ trim grid padding only after final decode
```

Hope this is useful for the next fix!

**Comment by Saganaki22** [comment]
update to v1.3.2 and test
